src/core/mavsdk_impl.cpp

```diff
@@ -73,49 +80,63 @@ std::string MavsdkImpl::version() const
     }
 
+    // TODO kalyan yeah this needs to be scrapped
     // FIXME: Ignore messages from QGroundControl for now. Usually QGC identifies
     //        itself with sysid 255.
     //        A better way would probably be to parse the heartbeat message and
     //        look at type and check if it is MAV_TYPE_GCS.
-    if (message.sysid == 255) {
-        return;
-    }
+    //if (message.sysid == 255) {
+        //return;
+    //}
 
     std::lock_guard<std::recursive_mutex> lock(_systems_mutex);
 
+    // TODO kalyan - I'm assuming this is how we create systems, and should be checked
     // Change system id of null system
-    if (_systems.find(0) != _systems.end()) {
-        auto null_system = _systems[0];
-        _systems.erase(0);
-        null_system->system_impl()->set_system_id(message.sysid);
-        _systems.insert(system_entry_t(message.sysid, null_system));
-    } else if (_is_single_system) {
-        auto sys = _systems.begin();
-        if (sys->first != message.sysid) {
-            sys->second->system_impl()->set_system_id(message.sysid);
-            _systems.insert(system_entry_t(message.sysid, sys->second));
-            _systems.erase(sys->first);
-        }
-    }
 
-    if (!does_system_exist(message.sysid)) {
-        make_system_with_component(message.sysid, message.compid);
-    } else {
-        _systems.at(message.sysid)->system_impl()->add_new_component(message.compid);
+    if (message.msgid == MAVLINK_MSG_ID_HEARTBEAT) {
+        std::cout << "Heartbeat: " << std::endl;
+        process_heartbeat(message);
     }
 
+    //if (_systems.find(0) != _systems.end()) {
+        //auto null_system = _systems[0];
+        //_systems.erase(0);
+        //null_system->system_impl()->set_system_id(message.sysid);
+        //_systems.insert(system_entry_t(message.sysid, null_system));
+    //} else if (_is_single_system) {
+        //auto sys = _systems.begin();
+        //if (sys->first != message.sysid) {
+            //sys->second->system_impl()->set_system_id(message.sysid);
+            //_systems.insert(system_entry_t(message.sysid, sys->second));
+            //_systems.erase(sys->first);
+        //}
+    //}
+
+    //if (!does_system_exist(message.sysid)) {
+        //make_system_with_component(message.sysid, message.compid);
+    //} else {
+        //_systems.at(message.sysid)->system_impl()->add_new_component(message.compid);
+    //}
+
     if (_should_exit) {
         // Don't try to call at() if systems have already been destroyed
         // in descructor.
         return;
     }
 
-    if (_systems.find(message.sysid) != _systems.end()) {
-        _systems.at(message.sysid)->system_impl()->process_mavlink_message(message);
+    //if (_systems.find(message.sysid) != _systems.end()) {
+        //_systems.at(message.sysid)->system_impl()->process_mavlink_message(message);
+    //}
+
+    uint16_t node_id = (message.sysid << 8) | message.compid;
+    if (_nodes.find(node_id) != _nodes.end()) {
```

Comment:
I would make this `if (node_exists(node_id)) {`, because I find that more readable than the `!= _nodes.end()` test. Later in the code, it is used again, and it could be `if (!node_exists(node_id))`. Again, just a suggestion.